### PR TITLE
Use same FOV definition within different reconstructions

### PR DESCRIPTION
**Pull request by TomTomRixRix**
So far all reconstruction algorithms implemented their own method to convert the field of view in mm into a field of view in pixels/voxels. To avoid errors, a single field of view conversion method is now used.

### CODE REVIEW

**Comment by TomTomRixRix** [comment]
Check if `flooring` or `rounding` in the compute image dimensions function makes more sense.

Visualize the field of view grid in the `create_a_custom_digital_device_twin.py` plot.

Add more tests to test specific spacings, shifted fovs and ranges.

**Comment by TomTomRixRix** [comment]
Rounding makes more sense as it will always lead to the smaller error compared to when always flooring. Important is that the sign of the difference needs to be considered in the following computation to adjust the start and end values accordingly (subtract in the case of flooring and add in the case of ceiling).

The field of view was added as an opaque red rectangle to the PA device plot.

Further tests with different spacings to cover more test cases were added.

**Comment by TomTomRixRix** [comment]
The tests currently fail due to issue #191

File: simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py
Comment (kdreher) [inline simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py:485]: I think we might run into an issue here in case we have the x.5 problem

File: simpa/core/processing_components/monospectral/field_of_view_cropping.py
Comment (kdreher) [inline simpa/core/processing_components/monospectral/field_of_view_cropping.py:51]: `int()` rounds down to an int, is this what we want here?

File: simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py
Comment (kdreher) [inline simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py:188]: see above

**Review by kdreher — COMMENTED** [review]
We might need to overthink the rounding in simpa

File: simpa/core/processing_components/monospectral/field_of_view_cropping.py
Comment (TomTomRixRix) [inline simpa/core/processing_components/monospectral/field_of_view_cropping.py:51]: I'm not sure if we want to round here at all or if we just want to truncate/cast to int. The `compute_image_dimensions` function converts the FOV in mm into pixel indices. The first three returned elements are the x, y and z dimension widths as integer values. The other `{x,y,z}dim_{start,end}` values are floats. Within `compute_image_dimensions` these floats are adjusted for offsets resulting from the rounding of the width to an integer. Applying an additional rounding here would then maybe make the FOV slightly larger than wanted, right? 

I'm not sure what the best way to convert the floats from `compute_image_dimensions` into ints is. When rounding away from zero or truncating/casting to int then the difference between the resulting `dim_start` and `dim_end` values might not be the same as the `dim_width` anymore. To achieve this, flooring or ceiling could be used but then the image might not be symmetric anymore. Depending on which behavior we would like to have we could use one or the other.

File: simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py
Comment (TomTomRixRix) [inline simpa/core/simulation_modules/reconstruction_module/reconstruction_utils.py:487]: This change only affects cases where `dim_temp` is an X.5 value. As the rounding difference (`diff`) is anyway compensated for in lines 489 and 490, `start` and `end` should only be affected in edge cases and there the maximal deviation should amount to the same as before but in the different direction. For `dim` it might, however, make an off-by-one difference, if e.g. 18.5 will be rounded to 18 or 19 so here the change to `round_away from_zero` makes sense 👍

File: simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py
Comment (TomTomRixRix) [inline simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py:59]: Do we then still need the 1e-10 here?

File: simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py
Comment (TomTomRixRix) [inline simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py:188]: same as above in the `field_of_view_cropping.py`

File: simpa_examples/create_a_custom_digital_device_twin.py
Comment (TomTomRixRix) [inline simpa_examples/create_a_custom_digital_device_twin.py:49]: What is the purpose of the `+50/2` ? Now the rectangle of the FOV is 25 pixels larger to the left than the sensor array

**Review by TomTomRixRix — COMMENTED** [review]
Overall, for all use cases of `np.round(...)` and `int(np.round(...))` we should switch to the new rounding function to avoid the X.5 issue. In cases where we used `int(...)` to truncate the value towards zero, adding a rounding functionality is maybe not what we want 🤔

File: simpa/utils/calculate.py
Comment (kdreher) [reply simpa/utils/calculate.py:297]: Done

File: simpa/utils/calculate.py
Comment (kdreher) [reply simpa/utils/calculate.py:299]: done

File: simpa/utils/calculate.py
Comment (kdreher) [reply simpa/utils/calculate.py:304]: done

File: simpa/utils/calculate.py
Comment (kdreher) [reply simpa/utils/calculate.py:317]: done

File: simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py
Comment (kdreher) [reply simpa/core/simulation_modules/reconstruction_module/time_reversal_adapter.py:59]: good catch :)

File: simpa_examples/create_a_custom_digital_device_twin.py
Comment (kdreher) [reply simpa_examples/create_a_custom_digital_device_twin.py:49]: you're right, I corrected it

File: simpa/core/processing_components/monospectral/field_of_view_cropping.py
Comment (kdreher) [reply simpa/core/processing_components/monospectral/field_of_view_cropping.py:51]: I corrected it :)

**Comment by TomTomRixRix** [comment]
All changes look good to me :+1: 

When running the manual tests there are differences between the reference and the test results. Especially in kwave simulations and reconstructions. This is likely due to the FOV changes, but I don't know if they are wanted like this. Perhaps double-check this before merging @kdreher Here are the reference (top) and test (bottom) results from the invision simulation test. The value ranges of the reconstruction are very different.

InvisionSimulationTest_reference
InvisionSimulationTest_test


I can't approve the PR as I'm the original author, so you probably have to do it @kdreher

**Comment by TomTomRixRix** [comment]
Also two manual tests failed because the adapter names which changed in #360 were not updated in another new branch #311 where the additional flags were introduced. I've now quickly fixed this bug.

**Review by kdreher — APPROVED** [review]
Looks good

**Pull request merged**
